unidler.py

```diff
@@ -48,14 +83,126 @@ def respond(self, status, body=None, headers={}):
+    ingresses = build_ingress_lookup()
+
+    with ingress_for_host(hostname, ingresses) as ing:
+        with deployment_for_ingress(ing) as deployment:
+            restore_replicas(deployment)
+            unmark_idled(deployment)
+            enable_ingress(ing)
+
+    with ingress(UNIDLER, 'default', ingresses) as unidler_ingress:
+        remove_host_rule(hostname, unidler_ingress)
+
+
+def build_ingress_lookup():
+    ingresses = client.ExtensionsV1beta1Api().list_ingress_for_all_namespaces()
+    return dict(
+        ((ingress.metadata.name, ingress.metadata.namespace), ingress)
+        for ingress in ingresses.items)
+
+
+@contextlib.contextmanager
+def ingress_for_host(hostname, ingresses):
+    for ingress in ingresses.values():
+        name = ingress.metadata.name
+        namespace = ingress.metadata.namespace
+
+        if (name, namespace) == (UNIDLER, 'default'):
+            continue
+
+        for rule in ingress.spec.rules:
+            if rule.host == hostname:
+                log.debug(
+                    f'Found ingress for {hostname}: {name} '
+                    f'in namespace {namespace}')
+                yield ingress
+                return write_ingress_changes(ingress)
+
+    raise IngressNotFound(f'Ingress for host {hostname} not found')
+
+
+def write_ingress_changes(ingress):
+    log.debug(
+        f'Writing changes to ingress {ingress.metadata.name} '
+        f'in namespace {ingress.metadata.namespace}')
+    client.ExtensionsV1beta1Api().patch_namespaced_ingress(
+        ingress.metadata.name,
+        ingress.metadata.namespace,
+        ingress)
+
+
+@contextlib.contextmanager
+def deployment_for_ingress(ingress):
+    name = ingress.metadata.name
+    namespace = ingress.metadata.namespace
+
+    try:
+        deployment = client.AppsV1beta1Api().read_namespaced_deployment(
+            name,
+            namespace)
+        log.debug(f'Found deployment {name} in namespace {namespace}')
+
+        yield deployment
+
+        write_changes(deployment)
+
+    except kubernetes.client.rest.ApiException as error:
+        raise DeploymentNotFound(f'Deployment {name} not found in {namespace}')
+
+
+def write_changes(deployment):
+    log.debug(
+        f'Writing changes to deployment {deployment.metadata.name} '
+        f'in namespace {deployment.metadata.namespace}')
+    client.AppsV1beta1Api().replace_namespaced_deployment(
+        deployment.metadata.name,
+        deployment.metadata.namespace,
+        deployment)
+
+
+def restore_replicas(deployment):
+    idled_at, replicas = deployment.metadata.annotations[IDLED_AT].split(',')
+    log.debug(f'Restoring {replicas} replicas')
+    deployment.spec.replicas = int(replicas)
+
+
+def enable_ingress(ingress):
+    log.debug('Enabling ingress')
+    ingress.metadata.annotations[INGRESS_CLASS] = 'nginx'
+
+
+def unmark_idled(deployment):
+    log.debug('Removing idled annotation and label')
+    del deployment.metadata.labels[IDLED]
```

Comment:
I was about to comment on the use of `filter()` to remove a key from a dict but you already changed it :)

Reply:
Also, something to bear in mind here is that `del` would raise an exception if the dictionary doesn't contain the key, in this case it *should not happen* as that would mean that we're unidling something which
1. is not unidled
2. or we didn't idle it properly - but then why we're unidling it (see 1.)

Reply:
Yes, the code could be a bit more defensive. Technically, it shouldn't be possible to unidle a non-idled deployment, because the request would never come to the unidler and run this code.